### PR TITLE
Ported Impulse Hopper

### PR DESCRIPTION
**Pull request by EpicSquid**
Advanced redstone modes currently are not implemented, as they need to be ported and most likely implemented for all machines that use redstone modes. 
Added "Lock Output" functionality, which prevents the impulse hopper from working while there are items in the output slot

### CODE REVIEW

File: enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/BlockImpulseHopper.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+import net.minecraft.util.BlockRenderLayer;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.EnumParticleTypes;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.IBlockAccess;
+import net.minecraft.world.World;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+
+public class BlockImpulseHopper extends BlockEio<TileImpulseHopper> implements IEioGuiHandler.WithPos, ITileEntityProvider, IHaveRenderers,
```
[the remark this answers is not in the document]
Comment (EpicSquid) [reply enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/BlockImpulseHopper.java]: I didn't realise when I started that AbstractMachineBlock was in the correct hierarchy. Also, vacuum chest or dialler didn't use it either iirc

File: enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/BlockImpulseHopper.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  @SideOnly(Side.CLIENT)
+  public IRenderMapper.IBlockRenderMapper getBlockRenderMapper() {
+    return ImpulseRenderMapper.instance;
+  }
+
+  @Override
+  @Nullable
+  public TileEntity createNewTileEntity(@Nonnull World worldIn, int meta) {
```
[the remark this answers is not in the document]
Comment (EpicSquid) [reply enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/BlockImpulseHopper.java]: Okay I'll check it out

File: enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/ContainerImpulseHopper.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      this.setY(y);
+      this.setSlot(slot);
+      this.setStackSizeLimit(64);
+      this.setDisplayStdOverlay(true);
+      this.setUpdateServer(true);
+    }
+
+    @Override
+    public void putStack(@Nonnull ItemStack stack, int realsize) {
+      getTileEntity().setGhostSlotContents(getSlot(), stack, realsize);
```
[the remark this answers is not in the document]
Comment (EpicSquid) [reply enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/ContainerImpulseHopper.java]: It wasn't syncing the client side ghost slots without it, so you couldn't see what was in them. I'll double check that's still the case though.

File: enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/GuiImpulseHopper.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import crazypants.enderio.base.gui.RedstoneModeButton;
+import crazypants.enderio.base.machine.gui.CapPowerBar;
+import crazypants.enderio.base.machine.gui.GuiButtonIoConfig;
+import crazypants.enderio.base.machine.gui.GuiOverlayIoConfig;
+import crazypants.enderio.machines.lang.Lang;
+import net.minecraft.client.gui.GuiButton;
+import net.minecraft.client.renderer.GlStateManager;
+import net.minecraft.entity.player.InventoryPlayer;
+
+public class GuiImpulseHopper extends GuiContainerBaseEIO implements ImpulseHopperRemoteExec.GUI {
```
[the remark this answers is not in the document]
Comment (EpicSquid) [reply enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/GuiImpulseHopper.java]: Can do

File: enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/ContainerImpulseHopper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (HenryLoenwind) [reply enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/ContainerImpulseHopper.java]: might be worth to trace out why, how it should work: super() sends a packet to the server which sets the data in the TE, which is synced to the client every tick the GUI is open, and the ghost slots read the data directly from the TE. This additional setter should only make the process faster and therefore more responsive, not be needed.

Unless I overlooked something... ;)

File: enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/BlockImpulseHopper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (HenryLoenwind) [reply enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/BlockImpulseHopper.java]: yeah. they probably have no good reason not to anymore

File: enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/BlockImpulseHopper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (EpicSquid) [reply enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/BlockImpulseHopper.java]: This has to be overridden even when extending AbstractMachineBlock apparently

File: enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/ContainerImpulseHopper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (EpicSquid) [reply enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/ContainerImpulseHopper.java]: hunting it down now

[the remark this answers is not in the document]
Comment (EpicSquid) [reply enderio-machines/src/main/java/crazypants/enderio/machines/machine/ihopper/ContainerImpulseHopper.java]: and now it works fine. The joys of coding :)

**Pull request merged**